test/config.ts

```diff
@@ -0,0 +1,71 @@
+import {
+  ChromiumEnv,
+  FirefoxEnv,
+  WebKitEnv,
+  test,
+  setConfig,
+  PlaywrightOptions,
+  Config,
+  globalSetup,
+} from "@playwright/test"
+import * as crypto from "crypto"
+import path from "path"
+import { PASSWORD } from "./utils/constants"
+import * as wtfnode from "./utils/wtfnode"
+
+// Playwright doesn't like that ../src/node/util has an enum in it
+// so I had to copy hash in separately
+const hash = (str: string): string => {
+  return crypto.createHash("sha256").update(str).digest("hex")
+}
+
+const cookieToStore = {
+  sameSite: "Lax" as const,
+  name: "key",
+  value: hash(PASSWORD),
+  domain: "localhost",
+  path: "/",
+  expires: -1,
+  httpOnly: false,
+  secure: false,
+}
+
+globalSetup(async () => {
+  console.log("\n🚨 Running globalSetup for playwright end-to-end tests")
+  console.log("👋 Please hang tight...")
+
+  if (process.env.WTF_NODE) {
+    wtfnode.setup()
+  }
+
+  const storage = {
+    cookies: [cookieToStore],
+  }
+
+  // Save storage state and store as an env variable
+  // More info: https://playwright.dev/docs/auth?_highlight=authe#reuse-authentication-state
+  process.env.STORAGE = JSON.stringify(storage)
+  console.log("✅ globalSetup is now complete.")
+})
+
+const config: Config = {
+  testDir: path.join(__dirname, "e2e"), // Search for tests in this directory.
+  timeout: 30000, // Each test is given 30 seconds.
+  retries: 3, // Retry failing tests 2 times
+}
+
+if (process.env.CI) {
+  config.retries = 2
```

Comment:
I guess this is to handle flakiness in CI environments?

Reply:
Yup! I thought the comment on L54 would suffice:
https://github.com/cdr/code-server/pull/3133/files/3b1802c2357c7707e487287391d16e93fb9eb756#diff-8c108cf05702abdfb2e259659b8823f7c917fced47193febd0ab7f4a39134294R54

But I'll add another

Reply:
that comment describes _what_ the setting does but not _why_ we need it, though :) IMO comments should be about the _why_ because it gives context

Reply:
Oops...I guess I thought the flakiness covered the _why_ but maybe not. Working on that 😅